File: logic/src/main/java/a75f/io/logic/bo/building/system/DefaultSystem.java
package a75f.io.logic.bo.building.system;

/**
 * Created by samjithsadasivan on 1/8/19.
 */

import java.util.HashMap;

import a75f.io.api.haystack.CCUHsApi;
import a75f.io.api.haystack.Equip;
import a75f.io.api.haystack.Tags;
import a75f.io.logger.CcuLog;
import a75f.io.logic.L;
import a75f.io.logic.bo.building.definitions.ProfileType;
import a75f.io.logic.bo.haystack.device.ControlMote;

/**
 * Default System Profile does nothing.
 *
 */
public class DefaultSystem extends SystemProfile
{
    @Override
    public String getProfileName() {
        return "Default";
    }
    
    @Override
    public ProfileType getProfileType() {
        return ProfileType.SYSTEM_DEFAULT;
    }
    
    public DefaultSystem() {
      addSystemEquip();
    }
    @Override
    public void doSystemControl() {
        updateOutsideWeatherParams();
        //update default points for apps and portals to consume
        String systemStatus = getStatusMessage();
        String scheduleStatus =  "No Central equipment is connected.";
        if (!CCUHsApi.getInstance().readDefaultStrVal("system and status and message").equals(systemStatus))
        {
            CCUHsApi.getInstance().writeDefaultVal("system and status and message", systemStatus);
        }
        if (!CCUHsApi.getInstance().readDefaultStrVal("system and scheduleStatus").equals(scheduleStatus))
        {
            CCUHsApi.getInstance().writeDefaultVal("system and scheduleStatus", scheduleStatus);
        }
    }
    
    @Override
    public void addSystemEquip() {
        CCUHsApi hayStack = CCUHsApi.getInstance();
        HashMap equip = hayStack.read("equip and system");
        if (equip != null && equip.size() > 0) {
            if (!equip.get("profile").equals(ProfileType.SYSTEM_DEFAULT.name())) {
                hayStack.deleteEntityTree(equip.get("id").toString());
            } else {
                return;
            }
        }
        CcuLog.d(L.TAG_CCU_SYSTEM,"Add Default System Equip");
        HashMap siteMap = hayStack.read(Tags.SITE);
        String siteRef = (String) siteMap.get(Tags.ID);
        String siteDis = (String) siteMap.get("dis");
        Equip systemEquip= new Equip.Builder()
                                   .setSiteRef(siteRef)
                                   .setDisplayName(siteDis+"-SystemEquip")
                                   .setProfile(ProfileType.SYSTEM_DEFAULT.name())
                                   .addMarker("equip").addMarker("system").addMarker("default")
                                   .addMarker("equipHis")
                                   .setTz(siteMap.get("tz").toString())
                                   .build();
        String equipRef = hayStack.addEquip(systemEquip);
        addSystemTuners();
        addDefaultSystemPoints(siteRef, equipRef, siteDis+"-SystemEquip", siteMap.get("tz").toString());
        updateGatewayRef(equipRef);

        new ControlMote(equipRef);
        L.saveCCUState();
        CCUHsApi.getInstance().syncEntityTree();
    }
    
    @Override
    public boolean isCoolingAvailable() {
        return false;
    }
    
    @Override
    public boolean isHeatingAvailable() {
        return false;
    }
    
    @Override
    public boolean isCoolingActive(){
        return false;
    }
    
    @Override
    public boolean isHeatingActive(){
        return false;
    }
    
    @Override
    public void deleteSystemEquip() {
        HashMap equip = CCUHsApi.getInstance().read("equip and system");
        if (equip.get("profile").equals(ProfileType.SYSTEM_DEFAULT.name())) {
            CCUHsApi.getInstance().deleteEntityTree(equip.get("id").toString());
        }
    }
    
    @Override
    public String getStatusMessage()
    {
        return "System is in gateway mode.";
    }
}

File: logic/src/main/java/a75f/io/logic/bo/building/system/DefaultSystem.kt
package a75f.io.logic.bo.building.system

import a75f.io.api.haystack.CCUHsApi
import a75f.io.domain.api.DomainName
import a75f.io.domain.equips.DefaultSystemEquip
import a75f.io.domain.logic.DeviceBuilder
import a75f.io.domain.logic.DomainManager
import a75f.io.domain.logic.DomainManager.addSystemDomainEquip
import a75f.io.domain.logic.EntityMapper
import a75f.io.domain.logic.ProfileEquipBuilder
import a75f.io.domain.util.ModelLoader.getCMDeviceModel
import a75f.io.domain.util.ModelLoader.getDefaultSystemProfileModel
import a75f.io.logger.CcuLog
import a75f.io.logic.L
import a75f.io.logic.autocommission.AutoCommissioningUtil
import a75f.io.logic.bo.building.definitions.ProfileType
import a75f.io.logic.bo.util.DesiredTempDisplayMode
import io.seventyfivef.domainmodeler.client.type.SeventyFiveFDeviceDirective
import io.seventyfivef.domainmodeler.client.type.SeventyFiveFProfileDirective
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

/**
 * Created by samjithsadasivan on 1/8/19.
 */

/**
 * Default System Profile does nothing.
 *
 */
class DefaultSystem : SystemProfile() {
    override fun getProfileName(): String {
        return "Default"
    }

    override fun getProfileType(): ProfileType {
        return ProfileType.SYSTEM_DEFAULT
    }


    override fun doSystemControl() {
        val defaultSystemEquip = DefaultSystemEquip(systemEquipRef)
        updateOutsideWeatherParams()
        //update default points for apps and portals to consume
        val systemStatus = statusMessage
        val scheduleStatus = "No Central equipment is connected."
        if (defaultSystemEquip.equipStatusMessage.readDefaultStrVal() != systemStatus) {
            defaultSystemEquip.equipStatusMessage.writeDefaultVal(systemStatus)
        }

        if (defaultSystemEquip.equipScheduleStatus.readDefaultStrVal() != scheduleStatus) {
            defaultSystemEquip.equipScheduleStatus.writeDefaultVal(scheduleStatus)
            if (AutoCommissioningUtil.isAutoCommissioningStarted()) {
                CcuLog.i(
                    L.TAG_CCU_SCHEDULER,
                    "System page status(when no central device is available) - AutoCommissioning is Started"
                )
                defaultSystemEquip.equipScheduleStatus.writeDefaultVal("In Diagnostic Mode")
            }
        }
    }

    fun createDefaultSystemEquip(): DefaultSystem {
        val hayStack = CCUHsApi.getInstance()
        val equip = hayStack.readEntity("equip and system and not modbus and not connectModule")
        if (equip != null && equip.isNotEmpty()) {
            if (equip["profile"] != DomainName.defaultSystemEquip) {
                hayStack.deleteEntityTree(equip["id"].toString())
                deleteOAODamperEquip()
                deleteBypassDamperEquip()
                removeSystemEquipModbus()
                deleteSystemConnectModule()
                removeSystemEquipBacnet()
            } else {
                return this
            }
        }
        CoroutineScope(Dispatchers.IO).launch {
            val defaultSystemModel = getDefaultSystemProfileModel() as SeventyFiveFProfileDirective
            val deviceModel = getCMDeviceModel() as SeventyFiveFDeviceDirective
            val entityMapper = EntityMapper(defaultSystemModel)
            val deviceBuilder = DeviceBuilder(hayStack, entityMapper)
            val equipBuilder = ProfileEquipBuilder(hayStack)
            val profileConfiguration = DefaultSystemConfig(defaultSystemModel)

            val equipDis = hayStack.siteName + "-" + defaultSystemModel.name
            val equipId = equipBuilder.buildEquipAndPoints(
                profileConfiguration,
                defaultSystemModel,
                hayStack.site!!.id,
                equipDis
            )
            val cmDevice = hayStack.readEntity("cm and device")
            if (cmDevice.isNotEmpty()) {
                hayStack.deleteEntityTree(cmDevice["id"].toString())
            }

            val deviceDis = hayStack.siteName + "-" + deviceModel.name
            deviceBuilder.buildDeviceAndPoints(
                profileConfiguration,
                deviceModel,
                equipId,
                hayStack.site!!.id,
                deviceDis, null
            )

            CcuLog.d(L.TAG_CCU_SYSTEM, "Add Default System Equip")
            updateAhuRef(equipId)

            addSystemDomainEquip(hayStack)
            DomainManager.addCmBoardDevice(hayStack)
            L.saveCCUState()

            DesiredTempDisplayMode.setSystemModeForDefaultSystemProfile(CCUHsApi.getInstance())
        }
        return this
    }    override fun addSystemEquip() {
        CcuLog.d(L.TAG_CCU_SYSTEM, "Updated temp modes for default system")
        DesiredTempDisplayMode.setSystemModeForDefaultSystemProfile(CCUHsApi.getInstance())
    }

    override fun isCoolingAvailable(): Boolean {
        return false
    }

    override fun isHeatingAvailable(): Boolean {
        return false
    }

    override fun isCoolingActive(): Boolean {
        return false
    }

    override fun isHeatingActive(): Boolean {
        return false
    }

    override fun deleteSystemEquip() {
        val equip =
            CCUHsApi.getInstance().readEntity("system and equip and not modbus and not connectModule")
        if (equip["profile"] == DomainName.defaultSystemEquip) {
            CCUHsApi.getInstance().deleteEntityTree(equip["id"].toString())
        }
        deleteOAODamperEquip()
        deleteBypassDamperEquip()
        removeSystemEquipModbus()
        deleteSystemConnectModule()
        removeSystemEquipBacnet()
    }

    override fun getStatusMessage(): String {
        return "System is in gateway mode."
    }
}
